### PR TITLE
Validate ORG server for topology based DS

### PR DESCRIPTION
**Pull request by rimashah25**
## What does this PR (Pull Request) do?

- [x] This PR is not related to any Issue 

## Which Traffic Control components are affected by this PR?

- Traffic Ops

## What is the best way to verify this PR?

1. Remove ORG Server from a given DS's topology
2. Try and update a DS and it should not allow you to do it.

Also, run the api/v3 test cases.

## The following criteria are ALL met by this PR

- [x] This PR includes tests
- [x] I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

## Additional Information

API route hasn't changed. Hence the documentation is not updated.

### CODE REVIEW

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 1]
```diff
@@ -1009,3 +1010,49 @@ GROUP BY t.name, ds.topology
...
+	servers[cacheGroupName] = serverName
+	fmt.Println(servers)
+
+	if len(servers) > 0 {
+		cachegroups, sysErr := GetTopologyCachegroups(tx, dsTopology)
+		if sysErr != nil {
+			return nil, fmt.Errorf("validating %s servers in topology: %v", tc.OriginTypeName, sysErr), http.StatusInternalServerError
+		}
+		for _, cg := range cachegroups {
+			if _, ok := servers[cg]; ok {
```
Comment (rawlinp) [inline traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:1050]: This logic is slightly off. We need to iterate over all the ORG server cachegroups to make sure they're all in the topology cachegroups. If any of them are not, we should return an error message that includes _all_ the offending ORG servers/cachegroups.

I would recommend converting `cachegroups` into a `map[string]struct{}` (essentially a set), then:
```
for cg, s := range servers {
    if _, ok := cachegroups[cg]; !ok {
        // add to list of offending servers/cachegroups
    }
}
// use strings.Join() to merge the offending servers/cachegroups into a single error message
```

**Review by rawlinp — CHANGES_REQUESTED** [review]

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 2]
```diff
@@ -1009,3 +1010,49 @@ GROUP BY t.name, ds.topology
...
 }
+
+// CheckOriginServerInCacheGroupTopology checks if a DS has ORG server and if it does, to make sure the cachegroup is part of DS
+func CheckOriginServerInCacheGroupTopology(tx *sql.Tx, dsID int, dsTopology string) (error, error, int) {
+	// get servers and respective cachegroup name that have ORG type in a delivery service
+	q := `
+		SELECT s2.host_name, c.name 
+		FROM server s2
+			INNER JOIN deliveryservice_server ds ON ds.server = s2.id
+			inner join type t on t.id = s2.type
```
[the remark this answers is not in the document]
Comment (rimashah25) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:1021]: Done

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 3]
```diff
@@ -1009,3 +1010,49 @@ GROUP BY t.name, ds.topology
 	}
 	return dsType, reqCap, topology, true, nil
 }
+
+// CheckOriginServerInCacheGroupTopology checks if a DS has ORG server and if it does, to make sure the cachegroup is part of DS
+func CheckOriginServerInCacheGroupTopology(tx *sql.Tx, dsID int, dsTopology string) (error, error, int) {
+	// get servers and respective cachegroup name that have ORG type in a delivery service
+	q := `
+		SELECT s2.host_name, c.name
```
[the remark this answers is not in the document]
Comment (rimashah25) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:1018]: Done

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 4]
```diff
@@ -1009,3 +1010,49 @@ GROUP BY t.name, ds.topology
...
+	rows, err := tx.Query(q, dsID)
+	if err != nil {
+		return nil, fmt.Errorf("querying deliveryservice origin server: %s", err), http.StatusInternalServerError
+	}
+	for rows.Next() {
+		if err := rows.Scan(&serverName, &cacheGroupName); err != nil {
+			return nil, fmt.Errorf("querying deliveryservice origin server: %s", err), http.StatusInternalServerError
+		}
+	}
+	fmt.Println("CacheName:", cacheGroupName)
```
[the remark this answers is not in the document]
Comment (rimashah25) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:1038]: Yes, those are for debugging purpose. I will remove it, when I convert to an actual PR

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 5]
```diff
@@ -1009,3 +1010,49 @@ GROUP BY t.name, ds.topology
...
+// CheckOriginServerInCacheGroupTopology checks if a DS has ORG server and if it does, to make sure the cachegroup is part of DS
+func CheckOriginServerInCacheGroupTopology(tx *sql.Tx, dsID int, dsTopology string) (error, error, int) {
+	// get servers and respective cachegroup name that have ORG type in a delivery service
+	q := `
+		SELECT s2.host_name, c.name 
+		FROM server s2
+			INNER JOIN deliveryservice_server ds ON ds.server = s2.id
+			inner join type t on t.id = s2.type
+			INNER JOIN cachegroup c ON c.id = s2.cachegroup
+		WHERE ds.deliveryservice=$1 and t.name='ORG'
```
[the remark this answers is not in the document]
Comment (rimashah25) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:1023]: Done

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 6]
```diff
@@ -1009,3 +1010,49 @@ GROUP BY t.name, ds.topology
...
+		GROUP BY s2.host_name, c.name
```
[the remark this answers is not in the document]
Comment (rimashah25) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:1024]: Without the GROUP BY clause, I get errors and unable to update DS.
https://dba.stackexchange.com/questions/88988/postgres-error-column-must-appear-in-the-group-by-clause-or-be-used-in-an-aggre

[the remark this answers is not in the document]
Comment (rawlinp) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:1024]: What is the error exactly? When I run this query without the group by, I get no errors:
```
to_development=# SELECT s2.host_name, c.name
FROM server s2
    INNER JOIN deliveryservice_server ds ON ds.server = s2.id
    inner join type t on t.id = s2.type
    INNER JOIN cachegroup c ON c.id = s2.cachegroup
WHERE ds.deliveryservice=10276 and t.name='ORG';
 host_name |        name
-----------+---------------------
 cfg2      | multi-site-org-west
(1 row)
```

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go (hunk 1 above)
Comment (rimashah25) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go]: Ok.. Updated the logic.

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (rimashah25) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go]: This was the error I used to get 
trafficops_1       | ERROR: api.go:209: 2020-11-17T22:26:29.8697024Z: 172.16.239.1:47762 querying deliveryservice origin server: pq: column "c.name" must appear in the GROUP BY clause or be used in an aggregate function
trafficops_1       | DEBUG: api.go:214: 2020-11-17T22:26:29.869793Z: Internal Server Error

but now i see none.

**Review by rawlinp — CHANGES_REQUESTED** [review]

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 7]
```diff
@@ -1090,3 +1090,53 @@ GROUP BY t.name, ds.topology
...
+	`
+
+	serverName := ""
+	cacheGroupName := ""
+	servers := make(map[string]string)
+	var offendingSCG []string
+	rows, err := tx.Query(q, dsID, tc.OriginTypeName)
+	if err != nil {
+		return nil, fmt.Errorf("querying deliveryservice origin server: %s", err), http.StatusInternalServerError
+	}
```
[the remark this answers is not in the document]
Comment (rimashah25) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:1113]: Done.

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 8]
```diff
@@ -1090,3 +1090,53 @@ GROUP BY t.name, ds.topology
...
+		}
+		for cg, s := range servers {
+			_, ok := topoCachegroups[cg]
+			if !ok {
+				offendingSCG = append(offendingSCG, fmt.Sprintf("%s-%s", s, cg))
+			}
+		}
+	}
+	if len(offendingSCG) > 0 {
+		return errors.New("list of server-cachegroup: " + strings.Join(offendingSCG, ", ") + " not present in this DS"), nil, http.StatusBadRequest
```
[the remark this answers is not in the document]
Comment (rimashah25) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:1139]: Sure, changed.

File: traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go [hunk 9]
```diff
@@ -1090,3 +1090,53 @@ GROUP BY t.name, ds.topology
...
+		}
+		// put slice values into map
+		topoCachegroups := make(map[string]string)
+		for _, cg := range cachegroups {
+			topoCachegroups[cg] = ""
+		}
+		for cg, s := range servers {
+			_, ok := topoCachegroups[cg]
+			if !ok {
+				offendingSCG = append(offendingSCG, fmt.Sprintf("%s-%s", s, cg))
```
[the remark this answers is not in the document]
Comment (rimashah25) [reply traffic_ops/traffic_ops_golang/dbhelpers/db_helpers.go:1134]: ok, changed.

File: traffic_ops/testing/api/v3/deliveryservices_test.go [hunk 10]
```diff
@@ -855,6 +856,83 @@ func UpdateDeliveryServiceWithInvalidSliceRangeRequest(t *testing.T) {
 
 }
 
+// UpdateValidateORGServerCacheGroup validates ORG server's cachegroup are part of topology's cachegroup
+func UpdateValidateORGServerCacheGroup(t *testing.T) {
+	msoDS := testData.DeliveryServices[7]
```
[the remark this answers is not in the document]
Comment (rimashah25) [reply traffic_ops/testing/api/v3/deliveryservices_test.go:861]: Understood and corrected.

File: traffic_ops/testing/api/v3/deliveryservices_test.go [hunk 11]
```diff
@@ -855,6 +856,83 @@ func UpdateDeliveryServiceWithInvalidSliceRangeRequest(t *testing.T) {
...
+		t.Fatalf("couldn't find any topologies: %v", err)
+	}
+	var p []int
+	var origNodes []tc.TopologyNode
+	newNodes := []tc.TopologyNode{{Id: 0, Cachegroup: "topology-edge-cg-01", Parents: p, LastUpdated: nil}}
+	if *msoDS.Topology == resp.Name {
+		origNodes = resp.Nodes
+		resp.Nodes = newNodes
+	}
+	_, _, err = TOSession.UpdateTopology(*msoDS.Topology, *resp)
```
[the remark this answers is not in the document]
Comment (rimashah25) [reply traffic_ops/testing/api/v3/deliveryservices_test.go]: ok.

**Review by rawlinp — APPROVED** [review]
Code looks good, and the new test passes

**Comment by mitchell852** [comment]
just for my own understanding, if a DS has 3 org servers directly assigned then it can have a topology assigned that includes those 3 org servers.....but if you remove any of those org servers from the topology, the ds cannot be saved as it has an _invalid_ topology, right?

**Comment by rimashah25** [comment]
> just for my own understanding, if a DS has 3 org servers directly assigned then it can have a topology assigned that includes those 3 org servers.....but if you remove any of those org servers from the topology, the ds cannot be saved as it has an _invalid_ topology, right? you would basically have to either:
> a. remove the topology from the ds or
> b. go fix the topology to include all 3 org servers

Yes, I believe so.

**Comment by rawlinp** [comment]
> but if you remove any of those org servers from the topology

With the next (and final) part of this ORG server-based validation, removing those ORG server cachegroups from the topology would be prohibited.

**Pull request merged**